File: legacy/lib/data/database/daos/follow_up_dao.dart
import 'package:drift/drift.dart';
import '../app_database.dart';
import '../tables.dart';

part 'follow_up_dao.g.dart';

@DriftAccessor(tables: [FollowUps])
class FollowUpDao extends DatabaseAccessor<AppDatabase> with _$FollowUpDaoMixin {
  FollowUpDao(super.db);

  Stream<List<FollowUpEntity>> watchByCustomer(String customerId) =>
      (select(followUps)
            ..where((t) => t.customerId.equals(customerId))
            ..orderBy([(t) => OrderingTerm(expression: t.scheduledAt, mode: OrderingMode.desc)]))
          .watch();

  Stream<List<FollowUpEntity>> watchToday(DateTime today) {
    final tomorrow = today.add(const Duration(days: 1));
    return (select(followUps)
          ..where((t) =>
              t.scheduledAt.isBiggerOrEqualValue(today) &
              t.scheduledAt.isSmallerThanValue(tomorrow) &
              t.completed.equals(false))
          ..orderBy([(t) => OrderingTerm(expression: t.scheduledAt)]))
        .watch();
  }

  Stream<List<FollowUpEntity>> watchUpcoming(DateTime from) {
    return (select(followUps)
          ..where((t) => t.scheduledAt.isBiggerOrEqualValue(from) & t.completed.equals(false))
          ..orderBy([(t) => OrderingTerm(expression: t.scheduledAt)]))
        .watch();
  }

  Future<String> insertFollowUp(FollowUpsCompanion entry) async {
    final id = entry.id.present ? entry.id.value : _genId();
    await into(followUps).insert(entry.copyWith(id: Value(id)));
    return id;
  }

  Future<void> markCompleted(String id) async {
    await (update(followUps)..where((t) => t.id.equals(id)))
        .write(FollowUpsCompanion(completed: const Value(true), completedAt: Value(DateTime.now())));
  }

  String _genId() {
    final now = DateTime.now();
    return 'fu_${now.microsecondsSinceEpoch}';
  }
}

File: app/src/main/java/com/salesquest/sales_quest/data/dao/FollowUpDao.kt
package com.salesquest.sales_quest.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import com.salesquest.sales_quest.data.entity.FollowUpEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface FollowUpDao {

    @Query("SELECT * FROM follow_ups WHERE customerId = :customerId ORDER BY scheduledAt DESC")
    fun watchByCustomer(customerId: String): Flow<List<FollowUpEntity>>

    @Query(
        "SELECT * FROM follow_ups WHERE scheduledAt >= :start AND scheduledAt < :end AND completed = 0 " +
            "ORDER BY scheduledAt"
    )
    fun watchToday(start: Long, end: Long): Flow<List<FollowUpEntity>>

    @Query(
        "SELECT * FROM follow_ups WHERE scheduledAt >= :from AND completed = 0 ORDER BY scheduledAt"
    )
    fun watchUpcoming(from: Long): Flow<List<FollowUpEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertFollowUp(followUp: FollowUpEntity)

    @Query("SELECT * FROM follow_ups")
    suspend fun getAll(): List<FollowUpEntity>

    @Query("UPDATE follow_ups SET completed = 1, completedAt = :completedAt WHERE id = :id")
    suspend fun markCompleted(id: String, completedAt: Long)

    @Query("DELETE FROM follow_ups")
    suspend fun clearAll()
}
